File: mobile/iosApp/PhoneShared/Utilities/ReminderSchedule.swift
import Foundation

/// Schedule maths for general (weight/meal/sleep) reminders — which days a
/// reminder is due, and when its one daily time falls. Pure Foundation so it
/// compiles in the widget extension too.
///
/// Mirrors `src/lib/server/push/reminders.ts` on the web and the Kotlin
/// `ReminderSchedule` on Android. `weekdays` uses the server's Sun=0..Sat=6
/// numbering, same as `SupplementSchedule` and supplement `scheduleDays`.
enum ReminderSchedule {
    static func isDue(weekdays: [Int], on date: Date, calendar: Calendar = .current) -> Bool {
        weekdays.contains(calendar.component(.weekday, from: date) - 1)
    }

    static func isDue(_ reminder: Reminder, on date: Date, calendar: Calendar = .current) -> Bool {
        isDue(weekdays: reminder.weekdays, on: date, calendar: calendar)
    }

    /// Every due-day occurrence of the reminder's time strictly after `from`
    /// within the next `days` days, ascending. Disabled reminders and
    /// unparseable times yield nothing.
    static func occurrences(
        for reminder: Reminder,
        from: Date = Date(),
        days: Int,
        calendar: Calendar = .current
    ) -> [Date] {
        // `parseTime` is a plain HH:MM parser with no supplement-specific
        // behavior — reused rather than duplicated.
        guard reminder.enabled, let time = SupplementSchedule.parseTime(reminder.time) else { return [] }
        var result: [Date] = []
        let today = calendar.startOfDay(for: from)
        for offset in 0 ..< days {
            guard let day = calendar.date(byAdding: .day, value: offset, to: today),
                  isDue(reminder, on: day, calendar: calendar),
                  let fireDate = calendar.date(
                      bySettingHour: time.hour, minute: time.minute, second: 0, of: day
                  ),
                  fireDate > from
            else { continue }
            result.append(fireDate)
        }
        return result
    }
}

File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/util/ReminderSchedule.kt
package com.bissbilanz.util

import com.bissbilanz.api.generated.model.Reminder
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.isoDayNumber

/**
 * Schedule maths for general logging reminders (weight / meal / sleep). Mirrors
 * `dueGeneralReminders` in `src/lib/server/push/reminders.ts` on the web — the two must
 * agree. [weekdays] uses the same Sun=0..Sat=6 numbering as [SupplementSchedule], via
 * `isoDayNumber % 7`.
 */
object ReminderSchedule {
    /** How far ahead [nextOccurrence] will look before giving up. */
    const val DEFAULT_HORIZON_DAYS = 60

    fun isDueOn(
        weekdays: List<Int>,
        date: LocalDate,
    ): Boolean = weekdays.contains(date.dayOfWeek.isoDayNumber % 7)

    fun isDueOn(
        reminder: Reminder,
        date: LocalDate,
    ): Boolean = isDueOn(reminder.weekdays, date)

    /**
     * The first due-day occurrence of [reminder]'s time strictly after [from], or null
     * when the reminder is disabled, its time is unparseable, or nothing falls inside
     * [horizonDays].
     */
    fun nextOccurrence(
        reminder: Reminder,
        from: LocalDateTime,
        horizonDays: Int = DEFAULT_HORIZON_DAYS,
    ): LocalDateTime? {
        if (!reminder.enabled) return null
        val time = SupplementSchedule.parseReminderTime(reminder.time) ?: return null
        var day = from.date
        repeat(horizonDays + 1) {
            if (isDueOn(reminder.weekdays, day)) {
                val candidate = LocalDateTime(day, time)
                if (candidate > from) return candidate
            }
            day = day.plusDays(1)
        }
        return null
    }
}

private fun LocalDate.plusDays(days: Int): LocalDate = LocalDate.fromEpochDays(toEpochDays() + days)
